source/auth/auth.rst

```diff
@@ -568,6 +568,18 @@ mechanism
 mechanism_properties
 	MUST NOT be specified.
 
+Caching credentials
+`````````````````````
+
+Drivers MUST maintain a cache of computed SCRAM credentials, where each cache
+entry's key is the tuple of ``(mongo_hashed_password, salt, i)``. Drivers SHOULD
+also include the name of the authentication mechanism, i.e. "SCRAM-SHA-1", as
+part of the key; this allows a single cache to be used for all current and 
+future SCRAM authentication mechanisms that drivers may implement.
+
+The cache entry value MUST be either the ``saltedPassword`` parameter, the
+combination of the ``clientKey`` and ``serverKey`` parameters, or the ``storedKey``
+result.
```

Comment:
Where did `storedKey` enter consideration? AFAIK, the `clientKey` value is still needed to generate a client proof (see: [libmongoc](https://github.com/mongodb/mongo-c-driver/blob/1.9.3/src/mongoc/mongoc-scram.c#L386)), so merely having `storedKey` accessible in the cache wouldn't help recompute that value.

Disclaimer: @bjori implemented libmongoc's SCRAM-SHA-1 conversation, and I've only been working off of reading his original code. I'm not terribly familiar with the spec.

Assuming `storedKey` isn't needed, I might propose rewriting as "the cache entry value SHOULD contain the `saltedPassword` or the combination of `clientKey` and `serverKey` values, which are derived from `saltedPassword`.

Reply:
Let's stick with what the RFC discusses, ClientKey and ServerKey or just SaltedPassword.

Reply:
The storedKey should be listed as an optional parameter that MAY be cached along with clientKey (it's just the digested clientKey).  The really expensive thing is the saltedPassword, but if someone wants to cache all the constant HMACs derived from that, they might as well cache the storedKey, too so there are no crypto functions being called to generate constant factors.  It needs to be worded better.

I wouldn't attach too much value to the text of the RFC.  It's one of the worst structured/written RFCs I've ever read.

Reply:
I added the info about the `storedKey` due to it being mentioned in the ticket description. I'll remove it.